File: RoadRunnerJS/src/jsTest/kotlin/TestUtil.kt
import util.EPSILON
import kotlin.math.abs

object TestUtil {
    private fun numericalDerivative(x: List<Double>, ds: Double): List<Double> {
        val deriv = (0 until x.size - 2).map {
            (x[it + 2] - x[it]) / (2 * ds)
        }.toMutableList()
        deriv.add(0, deriv.first())
        deriv.add(deriv.last())
        return deriv
    }

    fun assertDerivEquals(x: List<Double>, dx: List<Double>, ds: Double, epsilon: Double, errorFreq: Double = 0.01): Boolean {
        val numDx = numericalDerivative(x, ds)
        val count = dx.zip(numDx)
            .map { abs(it.first - it.second) }
            .filter { it > epsilon }
            .count()
        val freq = count.toDouble() / x.size

        return freq <= errorFreq
    }

    fun assertContinuous(values: List<Double>, epsilon: Double): Boolean {
        return values.drop(1)
            .zip(values.dropLast(1))
            .map { it.first - it.second }
            .maxOrNull() ?: 0.0 < EPSILON
    }
}

File: src/roadrunnerjs/util/TestUtil.ts
import { insert } from "./ArrayUtil";

function numericalDerivative(x: number[], ds: number): number[] {
  let deriv = new Array(x.length - 2).map(
    (e, i) => (x[i + 2] - x[i]) / (2 * ds)
  );
  deriv = insert(deriv, 0, deriv[0]);
  deriv.push(deriv[deriv.length - 1]);
  return deriv;
}

export function expectDerivEquals(
  x: number[],
  dx: number[],
  ds: number,
  epsilon: number,
  errorFreq: number = 0.01
) {
  const numDx = numericalDerivative(x, ds);
  const count = dx
    .map((e, i) => Math.abs(e - numDx[i]))
    .filter((e) => e > epsilon).length;
  const freq = count / x.length;

  expect(freq).toBeLessThanOrEqual(errorFreq);
}

export function expectContinuous(values: number[], epsilon: number) {
  if (values.length < 2) {
    expect(true).toBe(true);
  } else {
    const valuesWithoutLast = values.slice(0, -1);
    const deltas = values.slice(1).map((e, i) => e - valuesWithoutLast[i]);

    expect(Math.max(...deltas)).toBeLessThan(epsilon);
  }
}
